### PR TITLE
Canonicalize accelerator names in Resources

### LINKED ISSUE #995: UX/backend: Lower-case GPU succeeds in launching, but block forever in exec
Adapted from a programmatic use case from Erick. Here's a CLI repro:
```
# Succeeds. Because we allow lower-case GPUs in launching.
sky launch -c myclus --gpus v100 ''

# This blocks forever. `sky queue` will show PENDING. 
# This is because [v100:1] does not fit [V100:1].
sky exec myclus --gpus v100 -- echo hi

# Works.
sky exec myclus --gpus V100 -- echo hi
```

An immediate fix is to canonicalize the gpu string during exec.

A bigger item is probably to programmatically check if some task requirement is not ever going to be satisfied, e.g., `exec --gpus some_other_gpu`, and immediately fail.

### PR DESCRIPTION
**Pull request by WoosukKwon**
Fixes #995.

To prevent potential bugs due to lower-case names, I think we should canonicalize the accelerator names in `Resources`. Because there are only < 20 accelerators (except TPUs) in three clouds, we can maintain a hard-coded list of accelerator names. Using them, we can avoid reading the whole catalogs in most cases.

Note that this list (`_ACCELERATORS`) must also include accelerators supported for local clusters.

Tested:
* `sky exec --gpus v100`, `sky exec --gpus V100`
* `sky exec --gpus a10g`, `sky exec --gpus A10g`, `sky exec --gpus A10G`

### CODE REVIEW

File: sky/utils/resources_utils.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+"""Resources related utilities."""
```
Comment (concretevitamin) [inline sky/utils/resources_utils.py:1]: We can call this module accelerator_registry? So it reminds people if we want to add a new accelerator, this is the module to change (see a related comment below).

File: sky/utils/resources_utils.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+from sky.clouds import service_catalog
+from sky.utils import ux_utils
+
+# List of all accelerators supported by SkyPilot.
+# NOTE: Must include accelerators supported for local clusters.
```
Comment (concretevitamin) [inline sky/utils/resources_utils.py:6]: Can you document what happens if:
- a name is in this list, but not in any catalog (can users launch it?)
- a name is not in this list, but in a catalog (can users launch it?)
- for developers, what actions, if any, should we take to keep this list and the catalogs in sync?

File: sky/utils/resources_utils.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    'Inferentia',
+    'K520',
+    'K80',
+    'M60',
+    'Radeon Pro V520',
+    'T4',
+    'T4g',
+    'V100',
+    'V100-32GB',
+    'Virtex UltraScale (VU9P)',
```
Comment (concretevitamin) [inline sky/utils/resources_utils.py:20]: For example, `show-gpus --all` doesn't show this name or L11, possibly others.

File: sky/utils/resources_utils.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
+"""Resources related utilities."""
+from sky.clouds import service_catalog
+from sky.utils import ux_utils
+
+# List of all accelerators supported by SkyPilot.
+# NOTE: Must include accelerators supported for local clusters.
+_ACCELERATORS = [
```
Comment (concretevitamin) [inline sky/utils/resources_utils.py:7]: Comment that these are the *canonicalized* accelerator names used internally. Users should call `canonicalize_accelerator_name()` to get them.

File: sky/utils/resources_utils.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+                                                 case_sensitive=False)
+    names = list(searched.keys())
+
+    # Exact match.
+    if accelerator in names:
+        return accelerator
+
+    if len(names) == 0:
+        with ux_utils.print_exception_no_traceback():
+            raise ValueError(f'Unknown accelerator: {accelerator}')
```
Comment (concretevitamin) [inline sky/utils/resources_utils.py:52]: UX: 
```
(py37) ~ » sky launch --gpus V1000 ''                                                              2 ↵
I 08-14 08:29:48 optimizer.py:865] No resource satisfying {'V1000': 1} on [AWS, Azure, GCP].
sky.exceptions.ResourcesUnavailableError: No launchable resource found for task sky-cmd. To fix: relax its resource requirements.
Hint: 'sky show-gpus --all' to list available accelerators.
      'sky check' to check the enabled clouds.

(py37) ~ » sky launch --gpus V1000 ''                                                              1 ↵
ValueError: Unknown accelerator: V1000
```
Top is master branch's error message, bottom is this branch.  I feel like the former message is more informative. Can we fall back to that somehow, or move that to here?

File: sky/utils/resources_utils.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+    if len(names) == 1:
+        return names[0]
+
+    if len(names) > 1:
+        with ux_utils.print_exception_no_traceback():
```
Comment (concretevitamin) [inline sky/utils/resources_utils.py:58]: Comment on an example that could trigger this?

File: sky/utils/resources_utils.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    """Returns the canonical accelerator name."""
+    # TPU names are always lowercase.
+    if accelerator.lower().startswith('tpu-'):
+        return accelerator.lower()
+
+    mapping = {name.lower(): name for name in _ACCELERATORS}
+    if accelerator.lower() in mapping:
+        return mapping[accelerator.lower()]
+
+    # A new accelerator can be added after upgrading SkyPilot.
```
Comment (concretevitamin) [inline sky/utils/resources_utils.py:40]: Is this meant to catch that  a user 
- just upgraded catalogs
- hasn't upgraded SkyPilot code so the hard-coded list above is still outdated?

**Review by concretevitamin — COMMENTED** [review]
This is awesome @WoosukKwon! Left some comments.

File: sky/utils/resources_utils.py (hunk 4 above)
Comment (WoosukKwon) [reply sky/utils/resources_utils.py:7]: Fixed.

File: sky/utils/resources_utils.py (hunk 7 above)
Comment (WoosukKwon) [reply sky/utils/resources_utils.py:40]: Yes. I clarified the comment. PTAL.

File: sky/utils/resources_utils.py (hunk 6 above)
Comment (WoosukKwon) [reply sky/utils/resources_utils.py:58]: Added a comment. This is unreachable for now.

File: sky/utils/resources_utils.py (hunk 5 above)
Comment (WoosukKwon) [reply sky/utils/resources_utils.py]: Good catch. The error will be handled by the optimizer.

File: sky/utils/resources_utils.py (hunk 2 above)
Comment (WoosukKwon) [reply sky/utils/resources_utils.py]: Added more comments. PTAL.

File: sky/utils/resources_utils.py (hunk 1 above)
Comment (WoosukKwon) [reply sky/utils/resources_utils.py]: Great point. Fixed!

**Comment by WoosukKwon** [comment]
Thanks for quick review @concretevitamin! Addressed all comments. PTAL.

BTW, I realized there is a potential problem in this approach. While this is NOT the case for our current catalogs, the clouds may add the same accelerator with different names (e.g., A10g and A10G) in the future. In such a case, its canonical name can be ambiguous. WDYT @concretevitamin? Should we also handle such cases?

File: sky/utils/accelerator_registry.py [hunk 8]
```diff
@@ -1,9 +1,21 @@
-"""Resources related utilities."""
+"""Accelerator registry."""
 from sky.clouds import service_catalog
 from sky.utils import ux_utils
 
-# List of all accelerators supported by SkyPilot.
+# Canonicalized names of all accelerators (except TPUs) supported by SkyPilot.
 # NOTE: Must include accelerators supported for local clusters.
+
```
Comment (concretevitamin) [inline sky/utils/accelerator_registry.py:7]: styleguide nit: L7 should have a `#`, and L18 should not be blank --  to form a block of comments from L5 to L17 for L19.

File: sky/utils/accelerator_registry.py [hunk 9]
```diff
@@ -1,9 +1,21 @@
...
+# 1. What if a name is in this list, but not in any catalog?
+# The name will be canonicalized, but the accelerator will not be supported.
+# Optimizer will print an error message.
+# 2. What if a name is not in this list, but in a catalog?
+# The name will be searched on the catalog with its case being ignored.
+# If a match is found, the name will be canonicalized to that in the catalog.
```
Comment (concretevitamin) [inline sky/utils/accelerator_registry.py:13]: In other words, is this list simply an optimization to short-circuit checking the catalog?  If so, should document here or point 3.

**Comment by concretevitamin** [comment]
> BTW, I realized there is a potential problem in this approach. While this is NOT the case for our current catalogs, the clouds may add the same accelerator with different names (e.g., A10g and A10G) in the future. In such a case, its canonical name can be ambiguous. WDYT @concretevitamin? Should we also handle such cases?

If this happens, we could define two such canonical names and require any user input to match them, possibly banning case insensitivity (which you have already done in the "unreachable case").   I doubt hardware makers would choose an ambiguous name, however.

**Review by concretevitamin — APPROVED** [review]
LGTM.  Reminder of this comment: https://github.com/skypilot-org/skypilot/pull/1075#discussion_r945306264

File: sky/utils/accelerator_registry.py (hunk 8 above)
Comment (WoosukKwon) [reply sky/utils/accelerator_registry.py:7]: Thanks for letting me know! Fixed.

File: sky/utils/accelerator_registry.py (hunk 9 above)
Comment (WoosukKwon) [reply sky/utils/accelerator_registry.py:13]: Yes it is. Fixed the comment.

**Comment by WoosukKwon** [comment]
I worked well on local clusters as well. @concretevitamin Can I merge this?

**Comment by concretevitamin** [comment]
LGTM - let's ship this

**Comment by WoosukKwon** [comment]
@istoica Could you add more? When do you think we might want to allow different names for the same accelerator?

**Comment by WoosukKwon** [comment]
> BTW, I realized there is a potential problem in this approach. While this
 is NOT the case for our current catalogs, the clouds may add the same
 accelerator with different names (e.g., A10g and A10G) in the future. In
 such a case, its canonical name can be ambiguous. WDYT @concretevitamin
 <https://github.com/concretevitamin>? Should we also handle such cases?

Oh now I understand the context. Actually, I later realized that my comment was wrong. We have another layer that converts our internal representation of the accelerators (i.e., the canonicalized names) to the accelerator names defined by each cloud. Thus, SkyPilot can handle the case that clouds have different names for the same accelerator.

**Pull request merged**